accepted/2023/sign-signature-provider-plugins.md

```diff
@@ -0,0 +1,524 @@
+                                                         'sha256', 'sha384', and 'sha512'. [default: SHA256]
+  -t, --timestamp-url <timestamp-url>                    RFC 3161 timestamp server URL. [default:
+                                                         http://timestamp.acs.microsoft.com/]
+  -td, --timestamp-digest <timestamp-digest>             Digest algorithm for the RFC 3161 timestamp server. Allowed
+                                                         values are sha256, sha384, and sha512. [default: SHA256]
+  -m, --max-concurrency <max-concurrency>                Maximum concurrency. [default: 4]
+  -v, --verbosity                                        Sets the verbosity level. Allowed values are 'none',
+  <Critical|Debug|Error|Information|None|Trace|Warning>  'critical', 'error', 'warning', 'information', 'debug', and
+                                                         'trace'. [default: Warning]
+  -?, -h, --help                                         Show help and usage information
+```
+
+The new command can then be used to sign artifacts.
+
+```text
+sign code azure-key-vault -kvu https://my.vault.azure.test/ -kvc MyCertificate -kvm -d Description -u http://description.test -b C:\ClassLibrary1\ClassLibrary1\bin\Debug\net8.0 ClassLibrary1.dll
+```
+
+### Sign artifacts using Windows certificate store
+
+First, the plugin must be installed.  The following command would download and install the latest version of the plugin.
+
+```text
+sign plugin install Microsoft.Windows.CertificateStore.Sign
+```
+
+The new command can then be used to sign artifacts.
+
+```text
+sign code certificate-store --store-location CurrentUser --store-name My --sha1fingerprint da39a3ee5e6b4b0d3255bfef95601890afd80709 -d Description -u http://description.test -b C:\ClassLibrary1\ClassLibrary1\bin\Debug\net8.0 ClassLibrary1.dll
+```
+
+## Requirements
+
+### Goals
+
+* Create a plugin model that enables pluggable signature providers.  A signature provider plugin will offer an alternate implementation of [`System.Security.Cryptography.AsymmetricAlgorithm`](https://learn.microsoft.com/dotnet/api/system.security.cryptography.asymmetricalgorithm) and the [`System.Security.Cryptography.X509Certificates.X509Certificate2`](https://learn.microsoft.com/dotnet/api/system.security.cryptography.x509certificates.x509certificate2), if available, corresponding to the private key.  Note that while there may be valid scenarios for signing with a raw asymmetric key pair, some signing operations require a certificate and will fail if a certificate is not available.
+* Make Sign CLI plugin-neutral.  While Sign CLI may install some core plugins (TBD), most plugins should be installed separately from Sign CLI itself, and Sign CLI's only interactions with any plugin should be through this plugin model.
+* Enable Sign CLI and plugins to version and release independently.
```

Comment:
What is the reason for this requirement?
Or rather, why do we think it's necessary to allow pretty much anybody to build a plugin? I understand that there are several providers which would want to add a plugin here, but is it really necessary to fully decouple them?
For example, why not build all of these in-repo and ask the providers to simply send a pull request?
At runtime the plugins could still be loaded dynamically, but you would avoid the complexity of acquisition, install and security.

Reply:
What's the reason to restrict the ability for anyone to develop and distribute those on their own? Delivery mechanisms exist, it's not like whole new infrastructure is built just for distributing plugins.
What if someone doesn't want their code to be open source or disagrees with Sign CLI license? What if someone doesn't want a third party to control the release of their code?

Reply:
> What if someone doesn't want their code to be open source or disagrees with Sign CLI license? What if someone doesn't want a third party to control the release of their code?

They can fork the project and build their own distribution of the tool with their own modifications.

Reply:
And then that fork would have to be maintained and we'll end up with several Sign CLI distributions with different signing capabilities. Why require extra gestures from third parties? The more friction is added to the process, the less desire there would be to participate.
Unless there are security concerns with plugin delivery and loading, efforts for adding plugins should be reduced.

Reply:
Sorry if I wasn't clear - I'm not saying we should not do it. My point was about cost/complexity versus benefit. The added complexity and cost to support separate distribution of plugins looks substantial. I was mostly looking for a discussion about the needs and if there are other ways to solve the problem.

There's a somewhat similar concept in the .NET SDK - workloads. The requirements there are obviously different, but it was (and still is) a substantial effort to develop and maintain this (and I don't think we even allow 3rd part workloads).

Also, the entire "dotnet tool" functionality is similar in some ways as well.

Reply:
>  the entire "dotnet tool" functionality is similar in some ways as well.

I had the same view. We shouldn't build this feature as spec'd in isolation of dotnet tools. They are (broad brush stroke) effectively identical. We should either do what Vitek is proposing (which is probably best) or build a first class add-in model that this tool can use.

The more I think about it, dotnet tools with source forking/sharing give you pretty much what the spec is proposing, and I think, via simpler mechanisms that are easier to understand and definitely cheaper to build.

Reply:
Removing the plugin load requirement would also allow this tool to be compiled with Native AOT.

That would be very useful for systems which either don't have/want a pre-installed .NET runtime, or don't allow JITing. I know there are some secure systems which don't allow JITing today, so I would not rule out that requirement in the future.

Reply:
Hi,

Trying to answer a few things here around the intents:

- The binary separation of the plugins was in part to enable any service capable of providing a signature the ability to do so via this tool. 
- We wanted one tool, not forks, to keep it as simple as possible for end users. Forks are hard to maintain and there's also the possibility that multiple providers could be needed for more complex scenarios down the road.
- We wanted to isolate a plugin's dependencies from the main tool. That could be parts of the Azure SDK for Key Vault, or could be the AWS SDK should AWS choose to write a plugin. Could be other things entirely. 
- We didn't want to be gatekeepers of who gets to play. That includes having to review PR's for additional providers. That doesn't really scale well.

Reply:
> We wanted to isolate a plugin's dependencies from the main tool. 

The proposed fork model does that even more naturally.

Reply:
There is a significant downside to forking -- maintenance. It is anticipated that we'll have at least 15-20 signature providers. For each of those to fork, which puts a significant burden on them to stay up to date with any changes and create new releases each time. That is likely more effort than many/most of those organizations providing the signatures would like to sign up for--it's a much easier lift for them to create a plugin once and update it if the interface changes than to continually release updated versions of an app. It also creates support issues for each of the providers as now they're front-line support for the app and not just their plugin.

If an app gets out of date, then that creates significant confusion for end users as to what features or bugs exist in which version of the app. This has already happened with private forks of the codebase to support other providers, where their version contains bugs already fixed in the main code. The support cost has also deterred public release of that fork.

Reply:
How often do you think that one of these signature providers would need to issue an update in the global tool model?